### PR TITLE
Take into account size as well in weight limiting.

### PR DESCRIPTION
**Pull request by eskimor**
To avoid oversized blocks as well. We already have weight limiting in place, by using the `proof_size` component of weight to track block usage we can use all the existing logic to limit the size of the block equally as time weights.

Also get rid of randomized selection in provisioner as all production chains already support prioritized selection. This simplifies the runtime code a bit and is no longer properly supported as of the code unification (enter/create_inherent) PR.

- [x] Tests
- [x] Run benchmarks
- [x] Sanity check weights

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3022054

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3040032

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3041305

**Comment by eskimor** [comment]
bot bench $ runtime polkadot runtime_parachains::paras_inherent
bot bench $ runtime westend runtime_parachains::paras_inherent
bot bench $ runtime rococo runtime_parachains::paras_inherent

**Comment by eskimor** [comment]
bot bench $ runtime westend runtime_parachains::paras_inherent
bot bench $ runtime rococo runtime_parachains::paras_inherent

**Comment by eskimor** [comment]
bot bench $ runtime rococo runtime_parachains::paras_inherent

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 1]
```diff
@@ -900,94 +941,41 @@ fn compute_entropy<T: Config>(parent_hash: T::Hash) -> [u8; 32] {
...
 	mut dispute_statement_set_valid: CheckValidityFn,
 	max_consumable_weight: Weight,
-	rng: &mut rand_chacha::ChaChaRng,
 ) -> (Vec<CheckedDisputeStatementSet>, Weight) {
 	// The total weight if all disputes would be included
-	let disputes_weight = multi_dispute_statement_sets_weight::<T, _, _>(&disputes);
+	let disputes_weight = multi_dispute_statement_sets_weight::<T>(&disputes);
 
 	if disputes_weight.any_gt(max_consumable_weight) {
+		log::debug!(target: LOG_TARGET, "Above mas consumable weight: {}/{}", disputes_weight, max_consumable_weight);
```
Comment (tdimitrov) [inline runtime/parachains/src/paras_inherent/mod.rs:960]:
```suggestion
		log::debug!(target: LOG_TARGET, "Above max consumable weight: {}/{}", disputes_weight, max_consumable_weight);
```

File: runtime/parachains/Cargo.toml [hunk 2]
```diff
@@ -60,7 +60,6 @@ sp-tracing = { git = "https://github.com/paritytech/substrate", branch = "master
 thousands = "0.2.0"
 assert_matches = "1"
 serde_json = "1.0.96"
-
```
Comment (slumber) [inline runtime/parachains/Cargo.toml:63]: accidentally deleted line?

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 3]
```diff
@@ -340,16 +340,46 @@ impl<T: Config> Pallet<T> {
...
+		// We are assuming (incorrectly) to have all the weight (for the mandatory class or even
+		// full block) available to us. This can lead to slightly overweight blocks, which still
+		// works as the dispatch class for `enter` is `Mandatory`. By using the `Mandatory`
+		// dispatch class, the upper layers impose no limit on the weight of this inherent, instead
+		// we limit ourselves and make sure to stay within reasonable bounds. It might make sense
+		// to subtract BlockWeights::base_block to reduce chances of becoming overweight.
+		let max_block_weight = {
+			let dispatch_class = DispatchClass::Mandatory;
+			let max_block_weight_full = <T as frame_system::Config>::BlockWeights::get();
+			log::debug!(target: LOG_TARGET, "Max block weight max block: {}", max_block_weight_full.max_block);
```
Comment (slumber) [inline runtime/parachains/src/paras_inherent/mod.rs:365]:
```suggestion
			log::debug!(target: LOG_TARGET, "Max block weight: {}", max_block_weight_full.max_block);
```
unless it's repeated intentionally?

**Review by tdimitrov — APPROVED** [review]

File: runtime/parachains/src/paras_inherent/mod.rs [hunk 4]
```diff
@@ -900,94 +941,41 @@ fn compute_entropy<T: Config>(parent_hash: T::Hash) -> [u8; 32] {
 /// 2. If exceeded:
 ///   1. Check validity of all dispute statements sequentially
 /// 2. If not exceeded:
-///   1. Sort the disputes based on locality and age, locality first.
-///   1. Split the array
-///   1. Prefer local ones over remote disputes
```
Comment (slumber) [inline runtime/parachains/src/paras_inherent/mod.rs:905]: Why do we no longer prioritize local disputes? It could also be done without randomness

**Review by slumber — APPROVED** [review]
Clean PR, thanks!

File: runtime/parachains/src/paras_inherent/mod.rs (hunk 4 above)
Comment (tdimitrov) [reply runtime/parachains/src/paras_inherent/mod.rs:905]: As far as I understood the dispute set is sorted (local before remote) so they are still prioritized.

**Comment by tdimitrov** [comment]
I did a burn in today. Runtime version: 0.9.41-e2eddbecf7a

Before running disputes (new images were fully deployed around 14:00 UTC):
without_disputes

After enabling disputes:
image

Additionally I saw a spike in relay chain block times:
Screenshot from 2023-07-13 21-02-26

And a finality lag around 60 blocks:
Screenshot from 2023-07-13 21-02-14

Also I noticed some errors in the logs:
```
15:12:24.721 ERROR tokio-runtime-worker parachain::candidate-backing: Failed to validate and make available: ValidationFailed(ValidationFailed("Check Validation Outputs: Bad request"))
```
I'm not sure if they are related to the test or to the past runtime upgrades.

**Comment by tdimitrov** [comment]
Oh no... I just realised I didn't upgrade the runtime... The results above contain only node side changes :angry:

**Comment by tdimitrov** [comment]
I've discussed the results with @ordian and @sandreim and understood that my expectations are not correct. Bigger block times and finality lag is expected during disputes and the values above are acceptable.

I ran another test today. What I did:
* Deployed master and enabled disputes to get some baseline results.
* Disabled disputes to make sure the network recovers :heavy_check_mark: 
* Reenabled disputes again and deployed this PR - no significant change in the results.
* Performed runtime upgrade with a binary from this PR - again no significant change in results.
* Disabled disputes to make sure the network recovers :heavy_check_mark: 

"The results":
image
image
image
image
image
image

Some timestamps:
* 07:45 - initiated disputes with master deployed
* 10:30 - PR deployed on all nodes
* 11:27 - Disputes stopped to check recovery
* 11:45 - Network fully recovered
* 11:53 - Enabled disputes again
* 12:33 - Runtime upgrade
* 13:15 - Disputes stopped (and network recovered)

There are some outliers in terms of block import times but the node count is not too much  - 2 to 6 nodes most of the time, one slot with around 50 nodes.

I think it's safe to merge.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3306312

**Comment by eskimor** [comment]
bot merge

**Pull request merged**